kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ServerTemplate.java

```diff
@@ -241,4 +243,48 @@ public int hashCode() {
         result = 31 * result + (capabilities != null ? capabilities.hashCode() : 0);
         return result;
     }
+
+    public ServerTemplate cloneServerTemplate() {
+        synchronized (lock) {
+            ServerTemplate serverTemplate = new ServerTemplate();
+            serverTemplate.setId(this.getId());
+            serverTemplate.setName(this.getName());
+            containersSpec.forEach(c -> serverTemplate.addContainerSpec(new ContainerSpec(c.getId(), c.getContainerName(), c.getServerTemplateKey(), c.getReleasedId(), c.getStatus(), cloneContainerConfigMap(c.getConfigs()))));
+
+            final Map<Capability, ServerConfig> cloneServerConfig = new EnumMap<>(Capability.class);
+            cloneServerConfig.putAll(configs);
+            serverTemplate.setConfigs(cloneServerConfig);
+
+            serverInstances.forEach(serverInstance -> serverTemplate.addServerInstance(new ServerInstanceKey(serverInstance.getServerTemplateId(), serverInstance.getServerName(), serverInstance.getServerInstanceId(), serverInstance.getUrl())));
+
+            final List<String> cloneCapabilities = new ArrayList<>();
+            cloneCapabilities.addAll(capabilities);
+            serverTemplate.setCapabilities(cloneCapabilities);
+
+            serverTemplate.setMode(mode);
+
+            return serverTemplate;
+        }
+    }
+
+    protected Map<Capability, ContainerConfig> cloneContainerConfigMap(final Map<Capability, ContainerConfig> configs) {
+        final Map<Capability, ContainerConfig> cloneConfigs = new EnumMap<>(Capability.class);
+        configs.entrySet().stream().forEach(entry -> cloneConfigs.put(entry.getKey(), cloneContainerConfig(entry.getValue())));
+
+        return cloneConfigs;
+    }
+
+    protected ContainerConfig cloneContainerConfig(ContainerConfig containerConfig) {
+        if (containerConfig instanceof RuleConfig) {
+            RuleConfig ruleConfig = (RuleConfig) containerConfig;
+            return new RuleConfig(ruleConfig.getPollInterval(), ruleConfig.getScannerStatus());
+        }
+
+        if (containerConfig instanceof ProcessConfig) {
+            ProcessConfig processConfig = (ProcessConfig) containerConfig;
+            return new ProcessConfig(processConfig.getRuntimeStrategy(), processConfig.getKBase(), processConfig.getKSession(), processConfig.getMergeMode());
+        }
+
+        return null;
```

Comment:
I think that this option is not valid, right? There is no other container config type apart from rule or process at this moment.
In such case I would recommend throwing here an exception. That way we make sure than if we add another container config type in the future then we won't forget to add it here.

Reply:
I think you are right, but we need new type Exception firstly, what do you think WrongContainerConfigTypeException? 
@cristianonicolai @jstastny-cz

Reply:
I don't think we need to have some specialized exception. This is not something what should be handled on application layer, rather we need to indicate that something is wrong in application, something what should be fixed by developer.

Reply:
done